### DISCUSSION TITLE
Feature request: Allow null-propagation operator for event accessors

### DISCUSSION THREAD

**Discussion by jnm2** [dotnet/csharplang#737] [General] [answered]
@SLaks commented on [Sun Mar 15 2015](https://github.com/dotnet/roslyn/issues/1276)

Why can't I write code like this?

``` C#
Process.GetProcessById(2)?.Exited += delegate { };
```

Error:

> CS0079   The event 'Process.Exited' can only appear on the left hand side of += or -=

This would be especially useful when writing property setters that add and remove event handlers to their values if non-null.

---

@DavidKarlas commented on [Thu Mar 26 2015](https://github.com/dotnet/roslyn/issues/1276#issuecomment-86364093)

Bump, Is this bug or?

---

@gafter commented on [Thu Mar 26 2015](https://github.com/dotnet/roslyn/issues/1276#issuecomment-86607388)

the ?. Operator never produces an lvalue, so this is by design.

---

@SLaks commented on [Thu Mar 26 2015](https://github.com/dotnet/roslyn/issues/1276#issuecomment-86625707)

Can this be reopened as a feature request?
This would be very useful for code like [this](https://github.com/SLaks/DroidMaster/blob/master/DroidMaster/Core/PersistentDevice.cs#L65-L66).

---

@DavidKarlas commented on [Thu Mar 26 2015](https://github.com/dotnet/roslyn/issues/1276#issuecomment-86627051)

Maybe add also properties setters into request like:

Object?.Prop = false;

I think it's very good feature request compared to some futuristic requests for C#7 ;)

---

@SLaks commented on [Tue Nov 15 2016](https://github.com/dotnet/roslyn/issues/1276#issuecomment-86649615)

<strike>Note that the delegate expression (eg, `x?.SomeEvent += ExpensiveFunction().SomeMethod`) would still be evaluated if `x` is null.</strike>

---

@series0ne commented on [Tue Sep 22 2015](https://github.com/dotnet/roslyn/issues/1276#issuecomment-142255075)

@SLaks you're not the only one that has raised this: 

http://stackoverflow.com/questions/32519200/c-sharp-6-0-null-propagation-operator-property-assignment/32715351#32715351

Congrats for also adding delegates/events to the mix; I never considered this!

---

@jnm2 commented on [Thu Sep 22 2016](https://github.com/dotnet/roslyn/issues/1276#issuecomment-248942045)

What would it take to not evaluate the RHS if the object is null? That would seem far more helpful at first blush.

It makes me happy that this is being considered. I've wished for this on a regular basis writing UI code.

---

@jnm2 commented on [Thu Sep 22 2016](https://github.com/dotnet/roslyn/issues/1276#issuecomment-248997853)

Just for context, I ran into wanting this again today.

---

@miloush commented on [Wed Oct 12 2016](https://github.com/dotnet/roslyn/issues/1276#issuecomment-253223534)

Just for the record, I expected this to work for attaching event handlers too, though I now understand what the issue is.

``` C#
object o = null;
INotifyCollectionChanged c = null;

c?.CollectionChanged += OnChanged;
(o as INotifyCollectionChanged)?.CollectionChanged += OnChanged;
((INotifyCollectionChanged)o)?.CollectionChanged += OnChanged;
```

@SLaks 

> Note that the delegate expression (eg, `x?.SomeEvent += ExpensiveFunction().SomeMethod`) would still be evaluated if `x` is null.

Not if it was rewritten to `if (x != null) x.SomeEvent += ...` or am I missing something? Do you want it to be evaluated?

---

@salmelo commented on [Wed Oct 19 2016](https://github.com/dotnet/roslyn/issues/1276#issuecomment-254977999)

Would just like to add a vote to this (and null conditional lvalues in general) for whatever that's worth.

I run into wanting this very frequently when writing code to bridge UI and internal objects.

Edit: Ideally I would want the right hand side not to be evaluated at all in the case of null, much as if you had written the `if` statement necessary to produce the desired effect now.

---

@SLaks commented on [Tue Nov 15 2016](https://github.com/dotnet/roslyn/issues/1276#issuecomment-260694426)

@miloush: You're right; it should check the LHS first.

---

@jnm2 commented on [Sat Apr 15 2017](https://github.com/dotnet/roslyn/issues/1276#issuecomment-294301046)

@SLaks Could you migrate this to https://github.com/dotnet/csharplang please?

---

@Misiu commented on [Wed Jun 07 2017](https://github.com/dotnet/roslyn/issues/1276#issuecomment-306785711)

Whats the status of this?

I'd like to do this:

`TabListControl?.SelectedIndexChanged += this.OnSelectedIndexChanged;`

but I must add extra check

```
if (TabListControl != null)
{
  TabListControl.SelectedIndexChanged += this.OnSelectedIndexChanged;
}
```

I've looked at _Null-coalescing assignments_, but this isn't something I could use in this specific case.

**Comment by dylanh724** [comment]
Left-hand null checkers would be both intuitive and wonderful. There's no reason in 2019 that null checking needs to be so bloaty for such a common task.

**Comment by Misiu** [comment]
Any updates on this? 🤞

**Comment by jnm2** [comment]
@CyrusNajmabadi decided to champion https://github.com/dotnet/csharplang/issues/2883 and believes that this falls under that scope as well as the other kinds of assignments (simple or compound) where the LHS contains a null-propagation operator.

**Comment by jnm2** [comment, marked as answer]
Currently part of https://github.com/dotnet/csharplang/issues/6045
